ledger/sandbox-on-x/src/test/suite/scala/com/daml/ledger/sandbox/bridge/validate/SequencerStateSpec.scala

```diff
@@ -0,0 +1,133 @@
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.sandbox.bridge.validate
+
+import com.codahale.metrics.MetricRegistry
+import com.daml.ledger.sandbox.bridge.BridgeMetrics
+import com.daml.ledger.sandbox.bridge.LedgerBridge.toOffset
+import com.daml.ledger.sandbox.bridge.validate.SequencerState.SequencerQueue
+import com.daml.lf.data.Ref
+import com.daml.lf.transaction.GlobalKey
+import com.daml.lf.value.Value
+import com.daml.lf.value.Value.ContractId
+import com.daml.metrics.Metrics
+import org.scalatest.flatspec.AnyFlatSpec
+import org.scalatest.matchers.should.Matchers
+
+import scala.util.chaining._
+
+class SequencerStateSpec extends AnyFlatSpec with Matchers {
+  private implicit val bridgeMetrics: BridgeMetrics = new BridgeMetrics(
+    new Metrics(new MetricRegistry)
+  )
+
+  "enqueue" should "update the sequencer state" in {
+    val empty = SequencerState()
+
+    val Seq(offset1, offset2, offset3) = (1L to 3L).map(toOffset)
+
+    val updatedKeys1 = Map(key(1) -> Some(cid(1)))
+    val updatedKeys2 = Map(key(2) -> Some(cid(2)))
+    val updatedKeys3 = Map(key(2) -> None)
+
+    val consumedContracts1 = Set(cid(0))
+    val consumedContracts2 = Set.empty[ContractId]
+    val consumedContracts3 = Set(cid(2))
+
+    val finalState = empty
+      .enqueue(offset1, updatedKeys1, consumedContracts1)
+      .enqueue(offset2, updatedKeys2, consumedContracts2)
+      .enqueue(offset3, updatedKeys3, consumedContracts3)
+
+    finalState shouldBe SequencerState(
+      sequencerQueue = Vector(
+        offset1 -> (updatedKeys1, consumedContracts1),
+        offset2 -> (updatedKeys2, consumedContracts2),
+        offset3 -> (updatedKeys3, consumedContracts3),
+      ),
+      keyState = Map(
+        key(1) -> (Some(cid(1)), offset1),
+        key(2) -> (None, offset3),
+      ),
+      consumedContractsState = Set(cid(0), cid(2)),
+    )
+  }
+
+  "enqueue" should "throw on offset not after its last ingested offset" in {
+    val populatedSequencerState =
+      SequencerState(
+        sequencerQueue = Vector(
+          toOffset(1L) -> (Map.empty, Set.empty),
+          toOffset(3L) -> (Map.empty, Set.empty),
+        ),
+        keyState = Map.empty,
+        consumedContractsState = Set.empty,
+      )
+
+    val act = (idx: Long) => populatedSequencerState.enqueue(toOffset(idx), Map.empty, Set.empty)
+    val expectedMessage = (idx: Long) =>
+      s"Offset to be enqueued (Offset(Bytes(000000000000000$idx))) is not higher than the last enqueued offset (Offset(Bytes(0000000000000003)))"
+
+    the[RuntimeException] thrownBy act(2L) should have message expectedMessage(2)
+    the[RuntimeException] thrownBy act(3L) should have message expectedMessage(3)
+  }
+
+  "dequeue" should "evict entries" in {
+    val Seq(offsetBefore, offset1, offset2, offset3, offset4, offsetAfter) =
+      (0L to 5L).map(toOffset)
+
+    val initialSequencerQueue: SequencerQueue = Vector(
```

Comment:
Isn't it brittle to create sequencer state by hand? You are running a risk of getting it wrong, not now but in a subsequent PR by someone else

Reply:
Slightly. I usually prefer conciseness, but fine, I'll build it idiomatically

Reply:
Done